go/libraries/doltcore/sqle/branch_control/binlog.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package branch_control
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sync"
+)
+
+const (
+	currentBinlogVersion = uint16(1)
+)
+
+//TODO: add stored procedure functions for modifying the binlog
+
+// Binlog is a running log file that tracks changes to tables within branch control. This is used for history purposes,
+// as well as transactional purposes through the use of the BinlogOverlay.
+type Binlog struct {
+	rows    []BinlogRow
+	RWMutex *sync.RWMutex
+}
+
+// BinlogRow is a row within the Binlog.
+type BinlogRow struct {
+	IsInsert    bool
+	Branch      string
+	User        string
+	Host        string
+	Permissions uint64
+}
+
+// BinlogOverlay enables transactional use cases over Binlog. Unlike a Binlog, a BinlogOverlay requires external
+// synchronization.
+type BinlogOverlay struct {
```

Comment:
I'm at least planning on using it for statement-level rollback, so that only valid operations succeed. Perhaps it won't extend to transactions for now.